workflow-ui/backstack-android/src/test/java/com/squareup/workflow1/ui/backstack/StateRegistryHolderTest.kt

```diff
@@ -0,0 +1,380 @@
+    val holderToSave = StateRegistryHolder(
+      onWillSave = {},
+      onRestored = {}
+    )
+    val parentToSave = SimpleStateRegistry().apply {
+      // Need to call restore before moving lifecycle state past INITIALIZED.
+      stateRegistryController.performRestore(null)
+    }
+    holderToSave.attachToParentRegistry("parentKey", parentToSave)
+    parentToSave.lifecycleRegistry.currentState = RESUMED
+
+    // Store some data in the system.
+    val childToSave = stateRegistryOf("key" to bundleOf("data" to "value"))
+    holderToSave.saveRegistryController("childKey", childToSave.stateRegistryController)
+
+    // Save the entire tree. This simulates what would happen just before a config change, e.g.
+    val parentSavedBundle = parentToSave.saveToBundle()
+
+    // Create a whole new tree, restored from our bundle.
+    val holderToRestore = StateRegistryHolder(
+      onWillSave = {},
+      onRestored = {}
+    )
+    val parentToRestore = SimpleStateRegistry().apply {
+      stateRegistryController.performRestore(parentSavedBundle)
+    }
+    holderToRestore.attachToParentRegistry("parentKey", parentToRestore)
+    parentToRestore.lifecycleRegistry.currentState = RESUMED
+    val childToRestore = SimpleStateRegistry()
+    holderToRestore.restoreRegistryControllerIfReady(
+      "childKey",
+      childToRestore.stateRegistryController
+    )
+    val restoredChildContent = childToRestore.savedStateRegistry.consumeRestoredStateForKey("key")!!
+
+    // Verify that our leaf data was restored.
+    assertThat(restoredChildContent.getString("data")).isEqualTo("value")
+  }
+
+  @Test fun `restores child state controller`() {
+    val holder = StateRegistryHolder(
+      onWillSave = {},
+      onRestored = {}
+    )
+    val parent = stateRegistryOf(
+      "parentKey" to bundleOf(
+        // The childKey is associated with a "saved" SavedStateRegistryController, so we need to
+        // give it the special internal bundle structure instead of just using bundleOf() directly.
+        "childKey" to stateRegistryOf(
+          "key" to bundleOf(
+            "data" to "value"
+          )
+        ).saveToBundle()
+      )
+    )
+
+    holder.attachToParentRegistry("parentKey", parent)
+    parent.lifecycleRegistry.handleLifecycleEvent(ON_CREATE)
+
+    val child = SimpleStateRegistry()
+    holder.restoreRegistryControllerIfReady("childKey", child.stateRegistryController)
+
+    val childState = child.savedStateRegistry.consumeRestoredStateForKey("key")!!
+    assertThat(childState.getString("data")).isEqualTo("value")
+  }
+
+  @Test fun `detach throws when called without attach`() {
+    val holder = StateRegistryHolder(
+      onWillSave = {},
+      onRestored = {}
+    )
+
+    val error = assertFailsWith<IllegalStateException> {
+      holder.detachFromParentRegistry()
+    }
+    assertThat(error).hasMessageThat().isEqualTo("Expected to be detached only when attached.")
+  }
+
+  @Test fun `save callback is invoked`() {
```

Comment:
nit: bonus points for specifying the condition under which it's invoked, since that's relevant to this test's coverage (and therefore to maintainers).